hazelcast/src/main/java/com/hazelcast/sql/impl/expression/math/PlusFunction.java

```diff
@@ -0,0 +1,105 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.math;
+
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.SqlErrorCode;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.BiExpressionWithType;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import com.hazelcast.sql.impl.type.QueryDataTypeFamily;
+
+import java.math.BigDecimal;
+
+/**
+ * Implements evaluation of SQL plus operator.
+ */
+public final class PlusFunction<T> extends BiExpressionWithType<T> implements IdentifiedDataSerializable {
+
+    public PlusFunction() {
+        // No-op.
+    }
+
+    private PlusFunction(Expression<?> operand1, Expression<?> operand2, QueryDataType resultType) {
+        super(operand1, operand2, resultType);
+    }
+
+    public static PlusFunction<?> create(Expression<?> operand1, Expression<?> operand2, QueryDataType resultType) {
+        return new PlusFunction<>(operand1, operand2, resultType);
+    }
+
+    @Override
+    public int getFactoryId() {
+        return SqlDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return SqlDataSerializerHook.EXPRESSION_PLUS;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T eval(Row row, ExpressionEvalContext context) {
+        Object left = operand1.eval(row, context);
+        if (left == null) {
+            return null;
+        }
+
+        Object right = operand2.eval(row, context);
+        if (right == null) {
+            return null;
+        }
+
+        QueryDataTypeFamily family = resultType.getTypeFamily();
+        if (family.isTemporal()) {
+            throw new UnsupportedOperationException("temporal types are unsupported currently");
+        }
+
+        return (T) evalNumeric((Number) left, (Number) right, family);
+    }
+
+    private static Object evalNumeric(Number left, Number right, QueryDataTypeFamily family) {
+        switch (family) {
+            case TINYINT:
+                return (byte) (left.byteValue() + right.byteValue());
+            case SMALLINT:
+                return (short) (left.shortValue() + right.shortValue());
+            case INT:
+                return left.intValue() + right.intValue();
+            case BIGINT:
+                try {
+                    return Math.addExact(left.longValue(), right.longValue());
+                } catch (ArithmeticException e) {
+                    throw QueryException.error(SqlErrorCode.DATA_EXCEPTION, "BIGINT overflow");
```

Comment:
I came across this error message (here and in many other expressions) when doing tests that causes overflow. The problem is that it is not specific - the user will just receive it in runtime, without any understanding of what exactly went wrong. 

When implementing new operators I do something like this:
```
throw QueryException.error(SqlErrorCode.DATA_EXCEPTION,
    "BIGINT overflow in ABS function (consider adding explicit CAST to DECIMAL)");
```
I borrowed this approach from Postgres error messages - they always try to provide better user experience by adding more context to the error message.

I propose to follow a similar approach in the expressions in this PR.

Reply:
fixed in f19f51f2f0627047116118bc6af56d018ffac613